body-tracking-samples/simple_3d_viewer/main.cpp

```diff
@@ -99,24 +103,172 @@ bool ParseInputSettingsFromArg(int argc, char** argv, InputSettings& inputSettin
         {
             inputSettings.CpuOnlyMode = true;
         }
+        else if (inputArg == std::string("OFFLINE"))
+        {
+            inputSettings.Offline = true;
+        }
+     
         else
         {
-            return false;
+            inputSettings.FileName = inputArg;
         }
+     
     }
     return true;
+
 }
 
-int main(int argc, char** argv)
-{
-    InputSettings inputSettings;
-    if (!ParseInputSettingsFromArg(argc, argv, inputSettings))
+
+void playFile(InputSettings inputSettings) {
+    //create the tracker and playback handle
+    k4a_calibration_t sensor_calibration;
+    k4abt_tracker_t tracker = NULL;
+    k4a_playback_t playback_handle = NULL;
+
+    const char* file = inputSettings.FileName.c_str();
+    if (k4a_playback_open(file, &playback_handle) != K4A_RESULT_SUCCEEDED)
     {
-        PrintUsage();
-        return -1;
+        printf("Failed to open recording\n");
+        return;
     }
-    PrintAppUsage();
 
+
+    if (k4a_playback_get_calibration(playback_handle, &sensor_calibration) != K4A_RESULT_SUCCEEDED)
+    {
+        printf("Failed to get calibration\n");
+        return;
+    }
+    ;
+
+    k4a_capture_t capture = NULL;
+    k4a_stream_result_t result = K4A_STREAM_RESULT_SUCCEEDED;
+
+    k4abt_tracker_configuration_t tracker_config = { K4ABT_SENSOR_ORIENTATION_DEFAULT };
+
+    tracker_config.processing_mode = inputSettings.CpuOnlyMode ? K4ABT_TRACKER_PROCESSING_MODE_CPU : K4ABT_TRACKER_PROCESSING_MODE_GPU;
+
+    VERIFY(k4abt_tracker_create(&sensor_calibration, tracker_config, &tracker), "Body tracker initialization failed!");
+
+    k4abt_tracker_set_temporal_smoothing(tracker, 1);
+
+    int depthWidth = sensor_calibration.depth_camera_calibration.resolution_width;
+    int depthHeight = sensor_calibration.depth_camera_calibration.resolution_height;
+
+    // Initialize the 3d window controller
+    Window3dWrapper window3d;
+
+    window3d.Create("3D Visualization", sensor_calibration);
+    window3d.SetCloseCallback(CloseCallback);
+    window3d.SetKeyCallback(ProcessKey);
+
+    while (result == K4A_STREAM_RESULT_SUCCEEDED)
+    {
+        result = k4a_playback_get_next_capture(playback_handle, &capture);
+        if (result == K4A_STREAM_RESULT_SUCCEEDED)
+        {
+            
+            //enque capture and pop results - asynchronous
+            k4a_wait_result_t queue_capture_result = k4abt_tracker_enqueue_capture(tracker, capture, K4A_WAIT_INFINITE);
```

Comment:
@yijiew does this capture need to be  protected to ensure it has a color and depth image or just a depth image?

Reply:
It doesn't take color image. It needs to be ensure it contains depth and IR image. 

---
In reply to: [402450643](https://github.com/microsoft/Azure-Kinect-Samples/pull/48#discussion_r402450643) [](ancestors = 402450643)

Reply:
So after the call to k4a_playback_get_next_capture and before the call to k4abt_tracker_enqueue_capture() you need to ensure there is an IR Image (If you have an IR image you are then guaranteed to have a depth image) but because you can't trust the recording file, you need to ensure it you not getting a capture with only color, or a stream with without depth.  If you do find the capture is missing the IR image, then I would drop that capture and get another rather than completely erroring out. Also an error message when dropping images be helpful for the user to detect they might have a bad recording.